umbim/wwan: DHCP and static configuration coexistence, bugfixes.

Leo-PL opened the pull request:
This bulilds on top of #2759 - it is included verbatim, and includes fixes inspired by #3680.

Main improvement is optional fallback to DHCP address configuration if static configuration from MBIM is not available, separately for IPv6 and IPv4, which is in line with MBIM specification 1.0. If IPv4-only or IPv6-only PDP type is selected, only sub-interface of the selected type is created.

Missing standard options from other protocols are included as well. With those fixes, MBIM finally works reasonably well with those fixes.

Run tested on TP-Link Archer C7v2 plus Sierra EM7455 (DW5811e MBIM flavour)

sch-m committed:
umbim: add support for non-dhcp mode

There are mbim compatible wwan modules available which do not support
the dhcp autoconfiguration. (e.g. gemalto Cinterion ELS81)

This adds the possibility to get the configuration parameters from mbim.

Signed-off-by: Martin Schiller <ms@dev.tdt.de>
Signed-off-by: Lech Perczak <lech.perczak@gmail.com>

Leo-PL committed:
umbim: use IP configuration provided by MBIM by default

Previously, DHCP was used. According to MBIM Specification v1.0 errata 1 [1],
section 10.5.20, MBIM_CID_IP_CONFIGURATION,
if MBIM information element containing IP configuration is available,
host shall use it, and fall back to in-band mechanisms to acquire it therwise -
therefore make static configuration the default.

[1] https://www.usb.org/document-library/mobile-broadband-interface-model-v10-errata-1-and-adopters-agreement

Signed-off-by: Lech Perczak <lech.perczak@gmail.com>

Leo-PL committed:
umbim: inherit "peerdns" option from parent interface

MBIM protocol handler should intherit "peerdns" options from parent
interface on sub-interfaces, otherwise upstream DNS servers are applied
regardless of configuration.

Signed-off-by: Lech Perczak <lech.perczak@gmail.com>

Leo-PL committed:
umbim: inherit firewall zone membership from parent interface

Fix an issue where subinterfaces were not added to the same
firewall zone as their parent.
Inspired by 64bb88841fb.

Signed-off-by: Lech Perczak <lech.perczak@gmail.com>

Leo-PL committed:
umbim: restructure IPv4/IPv6 handling

Check whether interface is configured per IP type, not per DHCP. This is
preparation to allow fallback to DHCP if static IP configuration is not
available, which is the default option for MBIM modems

Signed-off-by: Lech Perczak <lech.perczak@gmail.com>

Leo-PL committed:
umbim: use static config by default, fallback to DHCP

Finally, inspired by ModemManager's logic, make static configuration
obtained through MBIM control channel, preferred.
If IP configuration is not available this way, fallback to DHCP(v6) if
enabled, else do not create a sub-interface for unavailable IP type.

Signed-off-by: Lech Perczak <lech.perczak@gmail.com>

Leo-PL committed:
umbim: respect 'Enable IPv6 negotiation' option

Don't bring IPv6 part of interface up if it's disabled,
or system does not support it.

Signed-off-by: Lech Perczak <lech.perczak@gmail.com>

Leo-PL commented:
@adschm @bmork @sch-m @hyc could you take a look please?

hyc reviewed (commented):
This seems redundant with and maybe overrides the ipv6 option set in the the Luci config dialog.

hyc commented on package/network/utils/umbim/files/lib/netifd/proto/mbim.sh:
```diff
@@ -150,8 +150,10 @@ _proto_mbim_setup() {
+
 	[ "$pdptype" = "ipv4" -o "$pdptype" = "ipv4v6" ] && {
-		if [ -z "$dhcp" -o "$dhcp" = 0 ]; then
+		if [ -n "$ipv4address" ]; then
```
This is closer to the intent of my previous patch, looks good to me.
https://github.com/TDT-AG/openwrt/pull/1/commits/cacbab1605de31470da81e46348b3346d0fc054c

hyc commented on package/network/utils/umbim/files/lib/netifd/proto/mbim.sh:
```diff
@@ -17,6 +17,7 @@ proto_mbim_init_config() {
 	proto_config_add_string auth
 	proto_config_add_string username
 	proto_config_add_string password
+	[ -e /proc/sys/net/ipv6 ] && proto_config_add_string ipv6
```
I don't see any reason to check for system support here. Just honor whatever setting the user configured, as I did in https://github.com/TDT-AG/openwrt/pull/1/commits/76c27c2c0eab6f34e633517ca2a76ef916a2d67d

Leo-PL replied:
I disagree. There is no need to do any of ipv6-related stuff if it isn't built into the kernel. Same check is done in QMI protocol handler.

hyc replied:
OK, fair enough

Leo-PL commented:
> 
> 
> This seems redundant with and maybe overrides the ipv6 option set in the the Luci config dialog.

I don't see which line you mention here. Also, the protocol handler sets what is default, and LuCI has to align to it, not the other way.

sch-m commented:
Sorry for the late reply, I am very busy at the moment.

I have now looked at the whole thing again, and while testing it, I came across a problem that I also had earlier:

At an `ifdown` of a mbim wwan interface, which was initiated with static configuration, the **netifd crashes with a segfault**.
When using DHCP the error does not occur.

Apparently netifd has a problem with dynamic interfaces with proto == static.

Here is the gdb backtrace:
```
Core was generated by `/sbin/netifd'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x000055b2ad53aff8 in interface_data_flush (iface=0x55b2ad6e2d40) at interface.c:191
191             avl_for_each_element_safe(&iface->data, d, node, tmp)
(gdb) bt
#0  0x000055b2ad53aff8 in interface_data_flush (iface=0x55b2ad6e2d40) at interface.c:191
#1  0x000055b2ad53b44b in interface_flush_state (iface=0x55b2ad6e2d40) at interface.c:276
#2  0x000055b2ad53b5cb in __interface_set_down (iface=0x55b2ad6e2d40, force=true) at interface.c:329
#3  0x000055b2ad53bb3c in interface_set_available (iface=0x55b2ad6e2d40, new_state=false) at interface.c:486
#4  0x000055b2ad53b964 in interface_main_dev_cb (dep=0x55b2ad6e2e18, ev=DEV_EVENT_REMOVE) at interface.c:431
#5  0x000055b2ad54e71c in device_broadcast_cb (ctx=0x7fffdde1afbc, list=0x55b2ad6e2e18) at device.c:471
#6  0x00007f6d584888be in ?? ()
#7  0x0000002821719905 in ?? ()
#8  0x000055b2ad6e7d48 in ?? ()
#9  0x0000000000000000 in ?? ()
(gdb)
```

Maybe @nbd168 has an idea how to fix this.

@Leo-PL @hyc 
Can you confirm that with the static configuration there is a crash of netifd on an `ifdown`?

hyc commented:
I'm still running with #2759 plus my small patches (https://github.com/TDT-AG/openwrt/pull/1) and have seen no crashes. That patchset is already using dynamic interfaces so I would expect it to already show the problem if it was there. But I also have to point out I'm not running on a current vanilla OpenWRT release. I'm on TurrisOS 5.1.8 on my Turris Omnia, which I believe is based on OpenWRT 19.07.

Leo-PL commented:
@sch-m In production, I also backported the changes to 19.07 branch myself, where I do not observe the crash. When testing on master at b30924873062be47250e6b872347461c8221688d, I didn't observe this crash as well. Maybe it's due to some unrelated, recent change. I think I might have some time to test that tomorrow.

sch-m commented:
I just tested again with a very fresh vanilla openwrt master + the commits from this PR and get a segfault.
Hardware is a x86_64 system with Cinterion MV31-W.

The log output looks like this:
```
...
Fri Jul 30 06:37:55 2021 kern.info kernel: [   73.931166] netifd[2150]: segfault at 501 ip 0000000000418bcb sp 00007ffd1779aed0 error 4 in netifd[404000+1b000]
Fri Jul 30 06:37:55 2021 kern.info kernel: [   73.941885] Code: 55 53 48 89 fb 52 48 83 bf 70 01 00 00 00 74 0c 48 8d bf 48 01 00 00 e8 97 88 ff ff 48 8b ab f0 04 00 00 4c 8d ab f0 04 00 00 <4c> 8b 65 00 48 8b 83 f8 04 00 00 48 39 45 08 74 1e 48 89 ee 4c 89
...
```

Leo-PL commented:
I finally got a chance to test an IPv6-enabled SIM, and for some reason, my EM7455 failed to establish IPv6 PDN connection - immediately rejected with NwError 33. Both with OpenWrt + umbim and latest ModemManager. Strangely, Huawei E3372h did work with IPv6 on the very same SIM. So, no luck for proper testing of IPv6, I need to dig into the modem internals. By any chance, do you have access to IPv6-enabled carriers?

hyc commented:
I wish I did, but no. I think the major carriers in US do now.

bmork commented:
>  my EM7455 failed to establish IPv6 PDN connection - immediately rejected with NwError 33. Both with OpenWrt + umbim and latest ModemManager. Strangely, Huawei E3372h did work with IPv6 on the very same SIM.

That's odd. 33 is MBIM_NW_ERROR_REQUESTED_SERVICE_OPTION_NOT_SUBSCRIBED according to libmbim/mbim-enums.h, so I wouldn't expect it to work with any modem - assuming the APN etc is the same too.

I do have some dual stack APNs I can test with, but currently no OpenWrt device with an MBIM modem connected.  But maybe you could check if this manual procedure, which works for me on the EM7455 in my laptop, produces the same result?
```
root@miraculix:/tmp# umbim -d /dev/cdc-wdm0 -n subscriber
  readystate: 0001 - initialized
  simiccid: 89470305121011008195
  subscriberid: 242013050138190
root@miraculix:/tmp# umbim -d /dev/cdc-wdm0 -t 42 -n attach
  nwerror: 0000 - unknown
  packetservicestate: 0002 - attached
  uplinkspeed: 50000000
  downlinkspeed: 300000000
root@miraculix:/tmp# umbim -d /dev/cdc-wdm0 -t 42 -n connect ipv4v6:telenor.smart
  sessionid: 0
  activationstate: 0001 - activated
  voicecallstate: 0000 - none
  nwerror: 0000 - unknown
  iptype: 0003 - ipv4v6
root@miraculix:/tmp# umbim -d /dev/cdc-wdm0 -t 42 -n config
  sessionid: 0
  ipv4address: 10.153.153.108/29
  ipv4gateway: 10.153.153.109
  ipv4mtu: 1500
  ipv4dnsserver: 193.213.112.4
  ipv4dnsserver: 130.67.15.198
  ipv6address: 2a02:2121:345:21af:a410:a6e2:1d93:2d6f/64
  ipv6gateway: 2a02:2121:345:21af:b94a:c10:c6a9:b3cc
  ipv6mtu: 1540
  ipv6dnsserver: 2001:4600:4:fff:0:0:0:52
  ipv6dnsserver: 2001:4600:4:1fff:0:0:0:52
root@miraculix:/tmp# ifconfig wwan0 10.153.153.108/29 up
root@miraculix:/tmp# ip addr add 2a02:2121:345:21af:a410:a6e2:1d93:2d6f/64 dev wwan0
root@miraculix:/tmp# ip route add 8.8.8.8 dev wwan0
root@miraculix:/tmp# ip route add 2001:4860:4860::8888 dev wwan0
root@miraculix:/tmp# ping  8.8.8.8
PING 8.8.8.8 (8.8.8.8) 56(84) bytes of data.
64 bytes from 8.8.8.8: icmp_seq=1 ttl=115 time=125 ms
^C
--- 8.8.8.8 ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 124.745/124.745/124.745/0.000 ms
root@miraculix:/tmp# ping   2001:4860:4860::8888
PING 2001:4860:4860::8888(2001:4860:4860::8888) 56 data bytes
64 bytes from 2001:4860:4860::8888: icmp_seq=1 ttl=56 time=42.1 ms
^C
--- 2001:4860:4860::8888 ping statistics ---
2 packets transmitted, 1 received, 50% packet loss, time 1001ms
rtt min/avg/max/mdev = 42.130/42.130/42.130/0.000 ms
```

Note that there is one magic modem specific issue to be aware of when testing IPv6: The default bearer APN is often configured by one of the modem profiles, and the type must either match or be IPV4V6.  Make sure profile 1 is configured for IPV4V6. You can't do that using umbim only.  But you can do it woth AT commands.

I have this:

```
root@miraculix:/tmp# cat /dev/ttyUSB2 &
[2] 14578
root@miraculix:/tmp# echo -e "AT+CGDCONT?\r" >/dev/ttyUSB2
root@miraculix:/tmp# 
+CGDCONT: 1,"IPV4V6","telenor.smart","0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0",0,0,0,0
+CGDCONT: 2,"IPV6","ims","0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0",0,0,0,0

OK


```

Leo-PL commented:
@bmork thanks for the hint. When I manually set default bearer APN to IPv6, it did work. I wonder from where did this configuration come from - I'm using generic firmware, and network is Orange PL, which uses either plain IPv4 on "internet" APN or IPv6-only PDN + CLAT on "internetipv6" APN.
Here is the log:
```
root@router-test:~# umbim -d /dev/cdc-wdm0 -n subscriber
  readystate: 0001 - initialized
  simiccid: <redacted>
  subscriberid: <redacted>
root@router-test:~# umbim -d /dev/cdc-wdm0 -t 42 -n attach
  nwerror: 0000 - unknown
  packetservicestate: 0002 - attached
  uplinkspeed: 50000000
  downlinkspeed: 300000000
root@router-test:~# umbim -d /dev/cdc-wdm0 -t 42 -n connect ipv6:internetipv6
  sessionid: 0
  activationstate: 0001 - activated
  voicecallstate: 0000 - none
  nwerror: 0000 - unknown
  iptype: 0002 - ipv6
root@router-test:~# umbim -d /dev/cdc-wdm0 -t 42 -n config
  ipv6address: 2a00:f41:7075:f8dd:89c3:15d7:efc7:9362/64
  ipv6gateway: 2a00:f41:7075:f8dd:8951:7ff4:eadc:92ce
  ipv6mtu: 1500
  ipv6dnsserver: 2a01:1700:2:ffff:0:0:0:9f01
  ipv6dnsserver: 2a01:1700:3:ffff:0:0:0:9822
```
I had the default bearer set this way:
```
AT+CGDCONT?
+CGDCONT: 1,"IP","internet","0.0.0.0",0,0,0,0
```
And set it this way:
```
AT+CGDCONT?
+CGDCONT: 1,"IPV6","internetipv6","0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0",0,0,0,0
```

Using empty APN with IPV4V6 PDP type also does work.
And, not surprisingly, the MBIM protocol handler connected successfully as well.
Though, this dependency on default bearer is quite bad on Sierra's side, after all the modem supports, IIRC, 8 PDN sessions multiplexed through MBIM, so the modem should at least try to set up a dedicated bearer.

There is one caveat that I noticed: When using static configuration for IPv6, netifd doesn't setup 464xlat automatically, as it did for me on Huawei E3372h in NCM mode. I was quite surprised to see that work out of the box. When I configured it manually with default NAT64 prefix 64:ff9b::/96 however, it did work. In fact I'm posting this over this very link.

I see one more point for improvement: detect actual type of established connection. For my operator (OPL) it will always be IPv4 or IPv6, depending on APN name, and then set up only needed sub-interfaces based on that, and not just on settings in UCI. After testing this out, I can actually implement this. without risking breakage.

ptpt52 commented:
ready to get merged?

Leo-PL commented:
> ready to get merged?

Certainly ready for review, even without the improvement I mentioned above, which is "nice to have". Of course, I would certainly appreciate "Tested-by:" and "Reviewed-by:".

Leo-PL commented:
While I was studying policy routing for failover usecase, I stumbled upon commit e51aa699f7ca3ce83a0add622c0fd17d0caafc46, which ads passthrough of `ip4table` and `ip6table` parameters to sub-interfaces. I believe this should be also integrated to MBIM.

@dangowrt do you recall any particular reason, why this was added only for DHCP(v6) case for QMI, and not for static configuration?

Leo-PL committed:
umbim: pass ipXtable to child interfaces

Inspired by commti e51aa699f7ca, allow setting specific routing tables
via ip4table and ip6table options, by passing them on child interfaces
created by MBIM protocol handler.

Signed-off-by: Lech Perczak <lech.perczak@gmail.com>

sch-m commented:
> Edit: Nevermind, it seems that QMI does not create sub-interfaces for static configuration. Maybe MBIM could be refactored in the same way.

I have designed my patch in such a way that sub-interfaces are also created for the static configuration. I find it very irritating when depending on the protocol sometimes a sub-interface is created and sometimes not. I wanted the behavior to be always the same.

Leo-PL commented:
> > Edit: Nevermind, it seems that QMI does not create sub-interfaces for static configuration. Maybe MBIM could be refactored in the same way.
> 
> I have designed my patch in such a way that sub-interfaces are also created for the static configuration. I find it very irritating when depending on the protocol sometimes a sub-interface is created and sometimes not. I wanted the behavior to be always the same.

Makes sense, acknowledged.
Anyway, I intend no major changes to this PR - it's ready for testing and review. I use this on my routers already "in production".

ptpt52 commented:
Great! I am going to test on this.

ptpt52 commented:
IPv6 is not working properly, the lan port does not obtain an IPv6 address, and the client in the lan area does not obtain an IPv6 address

ptpt52 commented:
ip -6 route:

![11](https://user-images.githubusercontent.com/5331426/140199845-d4bee2e1-e103-46fe-a648-7cabdfd477ef.png)

Nowaker commented:
I'm very disappointed with how OpenWRT project is governed. First, @hyc submits multiple fixes #3317 that get rejected because "code wasn't separated" (even though it was totally broken and that single patch was required to get anything working with MC7455). While I didn't agree with the maintainers, I kept it civil and focused on the outcome - getting the code merged. Maintainers claimed "[legit changes went] go to the trash because @hyc closed this PR" and "Nobody rejected anything. You closed because you didn't want to wait, because of a comment from a third person." so I re-created this PR as a follow-up. Unfortunately, nothing progressed, it's been open for 11 months now. In the meantime, another PR #4405 shows up that's based on the code from this code. And I also just learned there's another PR #2759 that touches the subject. The community tries hard and submits good code but nothing gets reviewed or merged. :-( If there's no mental compute power from maintainers for reviews and guidance, I'd rather the maintainers just skimmed through the changes and accepted them right away. 1 minute review and merge, that's it. If the reviewing process doesn't work for certain parts of the project, change it. After all, the way UMBIM is now, it hasn't been functional for years. And the current process only ensures this won't change any time soon.

ptpt52 commented on package/network/utils/umbim/files/lib/netifd/proto/mbim.sh:
```diff
@@ -109,39 +127,103 @@ _proto_mbim_setup() {
+		fi
+	}
+
+	[ "$pdptype" = "ipv6" -o "$pdptype" = "ipv4v6" ] && {
+		if [ -n "$ipv6address" ]; then
+			json_init
+			json_add_string name "${interface}_6"
+			json_add_string ifname "@$interface"
+			json_add_string proto "static"
+			json_add_array ip6addr
+			json_add_string "" "$ipv6address"
+			json_close_array
+			json_add_string ip6gw "$ipv6gateway"
+			json_add_array dns
+			[ "$peerdns" = 0 ] || json_add_string "" "$ipv6dnsserver"
+			json_close_array
+			proto_add_dynamic_defaults
+			[ -n "$zone" ] && json_add_string zone "$zone"
+			[ -n "$ip6table" ] && json_add_string ip6table "$ip6table"
+			json_close_object
+			ubus call network add_dynamic "$(json_dump)"
+		elif [ "$dhcp" != 0 ]; then
+			echo "mbim[$$]" "Starting DHCPv6 on $ifname"
+			json_init
+			json_add_string name "${interface}_6"
+			json_add_string ifname "@$interface"
+			json_add_string proto "dhcpv6"
+			json_add_string extendprefix 1
+			proto_add_dynamic_defaults
+			[ -n "$zone" ] && json_add_string zone "$zone"
+			[ -n "$ip6table" ] && json_add_string ip6table "$ip6table"
+			json_close_object
+			ubus call network add_dynamic "$(json_dump)"
+		fi
```
I don't know how to change on this code to make IPv6 work.
but the code bellow works
```
        [ -n "$ip_6" ] && {
                echo "mbim[$$]" "Configure IPv6 on $ifname"
                ip=${ip_6%%/*}
                mask=${ip_6##*/}
                gateway=$(awk '$1=="ipv6gateway:" {print $2}' "$config")
                mtu=$(awk '$1=="ipv6mtu:" {print $2}' "$config")
                [ "$mtu" ] && ip -6 link set "$ifname" mtu "$mtu"

                proto_add_ipv6_address "$ip" "128"
                proto_add_ipv6_prefix "$ip_6"
                proto_add_ipv6_route "$gateway" "128"
                [ "$defaultroute" = 0 ] || proto_add_ipv6_route "::" 0 "$gateway" "$metric" "" "$ip_6"
                [ "$peerdns" = 0 ] || {
                        dns_servers=$(awk '$1=="ipv6dnsserver:" {printf "%s ",$2}' "$config")
                        for dns in $dns_servers; do
                                proto_add_dns_server "$dns"
                        done
                }   
        }  
```

ptpt52 replied:
ping @Leo-PL @sch-m @hyc

Leo-PL replied:
@ptpt52 I'm not sure what are you trying to accomplish. IPv6 worked for me both with static and DHCP configuration. There is no auto-switch between them by design.

Leo-PL commented:
@adschm Do you think, that sending this PR as patchset to mailing list could increase a chance of getting it merged? The code is running on one of my devices stable since I opened the PR, albeit with IPv4-only provider.

Leo-PL commented:
> 
> 
> IPv6 is not working properly, the lan port does not obtain an IPv6 address, and the client in the lan area does not obtain an IPv6 address

Do I understand correctly, that IPv6 prefix is not delegated to LAN? How long is the prefix you obtain from ISP?

Could you tell more about your connection? And possibly, gather logs from connection establishment? Maybe it requires static configuration (out-of-band through MBIM control protocol) or DHCP explicitly?

adschm commented:
@Leo-PL It appears that essentially nobody cares for these packages. And since you have mode the PR bigger, it will take longer to get everything proper and tested before you can ask for a review.

I don't think that anything will change fundamentally on the mailing list.

I'd have tried to highlight jow- here, but since the last comment implies it's not ready for merge, I don't think I should do that now.

Leo-PL commented:
Correct. I'll highlight you when I fix the IPv6 delegation thing, which seems to have never been implemented previously in upstream code.

Leo-PL committed:
umbim: log output of 'config' step

Display full configuration obtained using MBIM control channel in the
log, from umbim output verbatim, for easier troubleshooting, and in
preparation for parser refactoring.

Signed-off-by: Lech Perczak <lech.perczak@gmail.com>

Leo-PL committed:
umbim: add "_proto_mbim_get_field" helper

Add a new helper to extract IP configuration from umbim output. This is
required to extract fields which can possibly have multiple values,
namely IP addresses and DNS servers, and get rid of primitive parser
using 'eval' builtin without support for this.

Signed-off-by: Lech Perczak <lech.perczak@gmail.com>

Leo-PL committed:
umbim: support multiple-valued configuration fields

MBIM supports multiple values for IP address and DNS server, and such
configuration is available through output of MBIM. Use new helper
method to support adding multiple addresses and DNS servers to static
interfaces for both IPv4 and IPv6.

Signed-off-by: Lech Perczak <lech.perczak@gmail.com>

Leo-PL committed:
umbim: drop IP configuration parsing using 'eval'

Finally, when new helper is in use, drop old IP configuration parser.

Signed-off-by: Lech Perczak <lech.perczak@gmail.com>

Leo-PL committed:
umbim: delegate RFC7278 IPv6 prefixes from OOB config

Delegate prefixes received through MBIM control channel the same way, as
would be done through DHCP, according to RFC7278.

Signed-off-by: Lech Perczak <lech.perczak@gmail.com>

Leo-PL committed:
umbim: handle MTU configuration

Allow setting interface MTU through UCI. If this is not set,
use MBIM-provided MTU, if provided through control channel.
If separate MTUs are provided for IPv4 and IPv6, apply larger of them.
This is very unlikely and possible only for IPv4v6 dual-stack configuration.

Signed-off-by: Lech Perczak <lech.perczak@gmail.com>

Leo-PL committed:
umbim: extract common code from static and dhcp(v6) setup procedure

Beginnings and endings of sub-interface creation procedure were
literally duplicates - extract them outside if "if" blocks

Signed-off-by: Lech Perczak <lech.perczak@gmail.com>

Leo-PL committed:
umbim: include MBIM-provided DNS servers also with DHCP mode

In MBIM interfaces, DNS servers may be provided out-of-band regardless
whether DHCP is used for configuration, or not. Move the DNS
configuration outside "if" blocks to support that.

Signed-off-by: Lech Perczak <lech.perczak@gmail.com>

Leo-PL commented:
@sch-m @ptpt52 @Nowaker @hyc, could you retest? I fixed a couple of bugs, and rebased against current master as well. One thing remaining, is properly setting up a scoped default route, as is done in QMI, to avoid breaking multi-homed routers. 

I tried this:
```
diff --git a/package/network/utils/umbim/files/lib/netifd/proto/mbim.sh b/package/network/utils/umbim/files/lib/netifd/proto/mbim.sh
index 11c857961f..7bd2dcad5b 100755
--- a/package/network/utils/umbim/files/lib/netifd/proto/mbim.sh
+++ b/package/network/utils/umbim/files/lib/netifd/proto/mbim.sh
@@ -235,7 +235,21 @@ _proto_mbim_setup() {
                        done
                        json_close_array
 
-                       json_add_string ip6gw $(_proto_mbim_get_field ipv6gateway "$mbimconfig")
+                       json_add_array routes6
+                       local ipv6gateway=$(_proto_mbim_get_field ipv6gateway "$mbimconfig")
+                       json_add_object ""
+                       json_add_string target "$ipv6gateway"
+                       json_add_string netmask 128
+                       json_close_object
+                       [ "$defaultroute" = 0 ] || {
+                               json_add_object ""
+                               json_add_string target "::0"
+                               json_add_string netmask 0
+                               json_add_string gateway "$ipv6gateway"
+                               json_add_string source "$ipv6address"
+                               json_close_object
+                       }
+                       json_close_array
 
                        [ "$peerdns" = 0 ] || {
                                json_add_array dns
```
but routes do not appear in `ifstatus` output and are not created, of course. I think I need to convert this to use `proto_add_*` helpers, but I haven't yet figured out, how to connect that with creation of dynamic sub-interface for the static case, which I'd like to keep for sake of integrity.

This is going to be an absolute monstrosity to review, but now MBIM will finally become usable. Enough for today, time for rest.

hyc commented:
Not sure I'll be able to test against master since I'm on Turris OS 5 which is based on OpenWRT 19.07.

hyc commented:
Btw - if this PR is now too big to easily review, should consider merging the PRs it was built on, and handle things incrementally. Since none of this works in the current source, it won't make anything worse.

Leo-PL commented:
> 
> 
> Not sure I'll be able to test against master since I'm on Turris OS 5 which is based on OpenWRT 19.07.

@hyc For 19.07, you only need to cherry-pick two commits from master before applying this branch, to get MBIM to the same state as in master: 
umbim: fix whitespace issues (01f29a0c72e88e12c2c1a44bc8ac28c986fb8d35)
umbim: fix return value of proto_mbim_setup() (6bc4c0ae3eb3dff6e410b5b012ab0bec267adb5a)
See also https://github.com/Leo-PL/openwrt/commits/mbim-fixes_19.07 - N.B. I did not update that branch with current state of this PR, you'll have to rebase, but with the two above, should apply cleanly.

adschm commented:
> Btw - if this PR is now too big to easily review, should consider merging the PRs it was built on, and handle things incrementally. Since none of this works in the current source, it won't make anything worse.

@hyc The original PRs are unmergeable because you refused to make them proper. That's not an option.

hyc commented:
@adschm 
> The original PRs are unmergeable because you refused to make them proper. That's not an option.

I was referring to #2759 which the other PRs built on. My patches from the closed PR were also split up and available here https://github.com/TDT-AG/openwrt/pull/1. Of course just because the PR was closed doesn't matter - if anyone wanted the code they could still grab it. So that's a pretty lame excuse for leaving all this to bit rot.

Leo-PL commented:
Regardless if #2759 could be merged separately or not, it still needs to be reviewed. Either option is fine for me, if any chanages are needed for it, I can easily rebase. Also, it is likely, that I'll slightly reorder my commits a bit, to match "cleanup, bugfixes, new features" pattern. 

Let's keep it civil and refrain from discussing politics here - this doesn't get us any closer to fixing MBIM support right now.

If anyone has suggestions, on how to handle IPv6 default route setup properly - setting "ip6gw" option isn't one for this case - I'd be very welcome for them.

hyc commented:
> If anyone has suggestions, on how to handle IPv6 default route setup properly - setting "ip6gw" option isn't one for this case - I'd be very welcome for them.

Sorry I can't help here since my carrier doesn't do IPv6 I have no way to test.

Nowaker commented:
>  The original PRs are unmergeable because you refused to make them proper.

Apparently, "proper PR" is more important than proper working code in the repo. +1 for what @hyc said. But rather than pick up the rherotic from the maintainers, I'll repeat what I already said: If there's no mental compute power from maintainers for reviews and guidance, I'd rather the maintainers just skimmed through the changes and accepted them right away. 1 minute review and merge, that's it. If the reviewing process doesn't work for certain parts of the project, change it. After all, the way UMBIM is now, it hasn't been functional for years. And the current process only ensures this won't change any time soon.

adschm commented:
@Leo-PL Some of the wwan commit lack a commit message.

> Apparently, "proper PR" is more important than proper working code in the repo.

@Nowaker That's how it is. Did you ever wonder how those things "not working" are typically introduced? If you add some mess of changes, you will just have more to clean up later.

> If there's no mental compute power from maintainers for reviews and guidance, I'd rather the maintainers just skimmed through the changes and accepted them right away. 1 minute review and merge, that's it.

"1 minute" ... Doesn't really sound like you know what you are talking of.

>  After all, the way UMBIM is now, it hasn't been functional for years. 

Where are the reports? I haven't seen any "Fixes" remarks in the commit message(s). And if it's as broken as you say, I wonder why it's just the few people talking to each other here. I don't know how bad it is, but from the picture I get here it's a niche issue. And from the initial message, it sounds more like an improvement than a fix.

And if it's that important to you, why are there no Tested-by or Reviewed-by, at least for some of the changes? How could somebody merge this without having reviewed it if nobody else seems to be willing or capable? How would I know it's actually working if no Tested-by are here?

adschm commented:
@Leo-PL Having looked at it another time, I really don't think you are on a good path here. You have 25 commits on a package nobody wants to review - this will remove the slight remaining chance of getting enough Tested-by/Reviewed-by and convince somebody like me to merge it.

I'd strongly suggest to break it down into separate _PRs_. Clearly separate fixes and features. Encourage people to test and report. This might eventually allow to get at least some of the stuff in.

hyc commented:
>Where are the reports? I haven't seen any "Fixes" remarks in the commit message(s). If it's as broken as you say, I wonder why it's just the few people talking to each other here. I don't know how bad it is

@adschm considering there is no support at all in the admin UIs, it's not surprising that nobody else has been submitting bug reports for non-working features. Because nobody has been able to use it at all, and that has always been the whole point here. https://github.com/openwrt/luci/pull/4360

adschm commented:
> > Where are the reports? I haven't seen any "Fixes" remarks in the commit message(s). If it's as broken as you say, I wonder why it's just the few people talking to each other here. I don't know how bad it is
> 
> @adschm considering there is no support at all in the admin UIs, it's not surprising that nobody else has been submitting bug reports for non-working features. Because nobody has been able to use it at all, and that has always been the whole point here. [openwrt/luci#4360](https://github.com/openwrt/luci/pull/4360)

So it's not about bug-fixing at all, but just about adding the GUI support, i.e. a new feature?

hyc commented:
It is both. Because when you enable GUI support you find that the underlying scripts don't work, which is where this PR comes in.

Leo-PL commented:
> 
> 
> > The original PRs are unmergeable because you refused to make them proper.
> 
> Apparently, "proper PR" is more important than proper working code in the repo. +1 for what @hyc said. But rather than pick up the rherotic from the maintainers, I'll repeat what I already said: If there's no mental compute power from maintainers for reviews and guidance, I'd rather the maintainers just skimmed through the changes and accepted them right away. 1 minute review and merge, that's it. If the reviewing process doesn't work for certain parts of the project, change it. After all, the way UMBIM is now, it hasn't been functional for years. And the current process only ensures this won't change any time soon.

Actually, I'm with Adrian on this. In my day job, I'm an embedded Linux distro maintainer as well, and know the pains of this. Cut one corner here, another there, and you quickly end up with unmaintainable mess, which I'd like to avoid as well, even if this means maintaining 25 extra patches on top of official release myself for long time.
I do maintain some patches anyway, if I can upstream some of them - great, a few commits less to rebase every so often.

Code is here, up for grabs. Reviewing and testing it helps. That's what I ask for, not for bashing on maintainers. They owe us nothing, and have limited resources.

I once again ask for civil discussion and refraining from politics.


> 
> 
> It is both. Because when you enable GUI support you find that the underlying scripts don't work, which is where this PR comes in.

Yes, but working protocol support is essential, LuCI GUI is optional. My PR came from need to use MBIM modems natively without need to reconfigure them into other modes, etc.

> 
> 
> @Leo-PL Having looked at it another time, I really don't think you are on a good path here. You have 25 commits on a package nobody wants to review - this will remove the slight remaining chance of getting enough Tested-by/Reviewed-by and convince somebody like me to merge it.
> 
> I'd strongly suggest to break it down into separate _PRs_. Clearly separate fixes and features. Encourage people to test and report. This might eventually allow to get at least some of the stuff in.

Thank you for the idea - that is what I intend to do. I will probably take up _some_ changes from #2759 and refactor them a bit, to even more clearly split cleanup, bugfixes, and then new features, which can easily be added separately. 

Back on topic of IPv6 prefix delegation, @ptpt52, I'm looking forward to hearing from you.  I'd appreciate if you could test prefix delegation, and show the configuration used for testing, with logs from MBIM protocol handler. By any chance, do you use dual-stack IPv4v6 ISP? I don't have a slightest chance to test that scenario, as my test ISP (Orange PL) is either IPv4-only or IPv6-only+464XLAT.

ptpt52 commented:
> > IPv6 is not working properly, the lan port does not obtain an IPv6 address, and the client in the lan area does not obtain an IPv6 address
> 
> @ptpt52 Do I understand correctly, that IPv6 prefix is not delegated to LAN? How long is the prefix you obtain from ISP?
> 
> Could you tell more about your connection? And possibly, gather logs from connection establishment? Maybe it requires static configuration (out-of-band through MBIM control protocol) or DHCP explicitly?
> 
> Edit: I probably can reproduce this - I'll have to do some investigation - thanks for bringing this to my notice. In the same network (Orange PL) another device with Huawei E3372h in NCM mode works with IPv6 for me, and /64 is delegated to LAN.

@Leo-PL 
it seems there is not delegated prefix obtained from ISP
but it is more like a hack (maybe)
1. the IPv6 address is somewhat like `2408:8469:5720:6da:95a4:1cdf:96f4:9f11/64`
2. but I set it to the interface as `2408:8469:5720:6da:95a4:1cdf:96f4:9f11/128`
3. and then I use the `2408:8469:5720:6da:95a4:1cdf:96f4:9f11/64` as delegated prefix
4. the gateway is `2408:8469:5720:6da:c854:8e07:f4bc:f634`
5. setup route `{ "target": "2408:8469:5720:6da:c854:8e07:f4bc:f634", "netmask": "128" }`
6. setup route `{ "target": "::", "netmask": "0", "gateway": "2408:8469:5720:6da:c854:8e07:f4bc:f634", "metric": 20, "source": "2408:8469:5720:6da:95a4:1cdf:96f4:9f11\/64" } `

here is all the `json_dump `
```
{ "action": 0, "ifname": "wwan0", "link-up": true, "keep": true, "ipaddr": [ { "ipaddr": "10.53.129.52", "mask": "29" } ], "ip6addr": [ { "ipaddr": "2408:8469:5720:6da:95a4:1cdf:96f4:9f11", "mask": "128" } ], "routes": [ { "target": "0.0.0.0", "netmask": "0", "gateway": "10.53.129.53", "metric": 20, "source": "10.53.129.52\/29" } ], "routes6": [ { "target": "2408:8469:5720:6da:c854:8e07:f4bc:f634", "netmask": "128" }, { "target": "::", "netmask": "0", "gateway": "2408:8469:5720:6da:c854:8e07:f4bc:f634", "metric": 20, "source": "2408:8469:5720:6da:95a4:1cdf:96f4:9f11\/64" } ], "ip6prefix": [ "2408:8469:5720:6da:95a4:1cdf:96f4:9f11\/64" ], "dns": [ "119.7.7.7", "119.6.6.6", "2408:8001:7000:0:0:0:0:1", "2408:8001:7000:0:0:0:0:0" ], "interface": "usbwan" }
```

I am using the code from @hyc and with some changes on it
https://github.com/x-wrt/x-wrt/commit/d049bfad8b702996b9b12cea733de66586d3e240
+ @hyc 's changes:
https://github.com/x-wrt/x-wrt/commit/bd6162ddb8d37ac6a452764167a2c06b64d4e763

hyc commented:
> Code is here, up for grabs. Reviewing and testing it helps. That's what I ask for, not for bashing on maintainers. They owe us nothing, and have limited resources.

As a practical matter - if I was reviewing a patchset for modem connectivity, and it was broken up into individual patches each of which on its own is not enough to achieve connectivity, I'd be hard pressed to approve any of them. If the reviewers on this project aren't competent enough to read a few lines of shell script and evaluate its correctness, they need more competent reviewers.

Leo-PL commented:
@ptpt52 thank you, this actually makes total sense to me, and is what I'm trying to achieve. I see you're using IPv4v6, which I did not have available to test.
I Got similar IPv6 configuration from NCM modem in DHCP. Routes should be like that - one is link-local route - one with /128 mask and second is actual default route with scoped source address, to avoid creating blackhole for other IPv6 routable prefixes. This is explained here: https://openwrt-devel.openwrt.narkive.com/1Hpjoj2D/patch-uqmi-add-proper-ipv6-support#post2
I wonder why `ip6prefix` has the lower part set, despite being /64 - this would not happen if I created dynamic interface.

Leo-PL commented:
BTW. As practically all my changes touch only `mbim.sh`, for testing it is enough, to just copy it over `/lib/netifd/proto/mbim.sh`.

obsy commented:
Tested-by: Cezary Jackiewicz <cezary@eko.one.pl>

x86/64, DW5811e in mbim mode

Leo-PL commented:
Now that #4845 is merged, I will rebase this and resume work on fixing static IPv6 route setup, and refactor this a tiny bit.

Leo-PL commented:
@ptpt52 could you please test this in spare time? https://github.com/Leo-PL/openwrt/commit/9f7d6fdbed60631efe7bc791f1319c6513d8b668

I added scoped source routing, but not in the way I'd like to - please check if it works for you as well.
I'm seriously considering dropping the "use sub-interface for static configuration" requirement, as composing JSONs by hand is PITA, and I can't figure out a way to replace passing a scoped route through "ip6gw" option. As a last resort, I'll ask the forums.

Leo-PL commented:
Also, my other EM7455 managed to connect over DHCPv6 over MBIM, in addition to static configuration. No idea why the one I used originally, didn't.
I plan to make "dhcp" and "dhcpv6" options separate and more specific, as "off", "auto", and "force" to allow automatic selection between modes and forcing either.

Leo-PL committed:
umbim: connect session for only the selected PDP type

Previous implementation automatically set up connections for both IPv4
and IPv6, even if one of them isn't supported. Respect the "pdptype"
option in the same way, as it is done for QMI or NCM, and only start the
respective PDN sessions, if set.

Signed-off-by: Lech Perczak <lech.perczak@gmail.com>

Leo-PL committed:
umbim: detect actual connection IP type

Current implementation needlessly creates both IPv4 and IPv6
sub-interfaces for single-stack IP types. Limit this only to selected IP
type. While at that, ensure that IP type is also passed to umbim during
"connect" phase. In addition, detect the actual established connection
type returned by umbim and set up subinterfaces according to that,
not to requested configuration. While at that, allow empty IP type explicitly,
interpreted as "any" according to MBIM specification.

Signed-off-by: Lech Perczak <lech.perczak@gmail.com>

arinc9 commented:
@Leo-PL I've got someone with Quectel RM520N-GL, they get high latency, ~50ms, running on MBIM mode on the latest snapshot with luci-proto-modemmanager. They wanted to try luci-proto-mbim but I see a bunch of PRs so I was rather confused how to proceed. I believe this is the most recent PR. I intend to apply this and the luci-proto-mbim PR, https://github.com/openwrt/luci/pull/4360 to try it out. Let me know if you need any information.

Leo-PL commented:
@arinc9 Yeah, this is the most recent one. I intended to do some more refactoring, but first I need to rebase this, because of recent roaming support addition. Even if I only rebase - this can be merged fully or in part, provided that ordering of commits is maintained. Thanks for looking into this - I'll rebase this in the evening today.

Leo-PL commented:
@arinc9 it'll take a bit longer than I expected, because I found one commit which got squashed by accident, and it needs splitting again. I also think, that while doing the redesign, selection for DHCP and DHCPv6 shall be split before we merge it, as this affects the UCI properties, so it's especially important to get it right the first time.

Leo-PL committed:
umbim: fail connect step immediately

Subsequent calls to 'umbim connect' do not have any effect if a failure
occured, and in such case an infinite loop without timeout is created,
leading to possibility of interface stuck at connecting forever.
Drop this loop, and issue MBIM disconnect properly, so netifd can
restart from scratch.
This issue can be observed with Sierra EM7455 at changing APN, which
causes network re-registration by default, and a MBIM transaction
timeout, which is resolved on next interface bringup by netifd.

Signed-off-by: Lech Perczak <lech.perczak@gmail.com>

Leo-PL committed:
umbim: separate DHCPv6 configuration from DHCP(v4)

Now, that sub-interface setup is split by IP type, and separate checks
are performed for DHCP selection, it is possible to control DHCP on v4
an v6 sub-interfaces instantly. Add "dhcpv6" variable, akin to QMI
option, to control behaviour of DHCPv6 separately from IPv4 option,
which is required for some mobile operators.

Signed-off-by: Lech Perczak <lech.perczak@gmail.com>

arinc9 commented:
All good. I see some new changes, is it ready to test?

Leo-PL commented:
@arinc9 I haven't had a chance to retest IPv6 yet, but you can try it out anyway. I plan to implement the "off/auto/force" logic to DHCP(v6) selection, too. I'll mark the PR ready once I finish the planned changes.
LuCI protocol plugin will require some minor changes as well, due to added 'dhcpv6' flag in UCI.

arinc9 commented:
@Leo-PL Great, will @hyc's luci-proto-mbim PR work with this? Or are there changes required for their luci-proto-mbim?

Leo-PL commented:
IIRC only for the DHCP switching, which I modified only yesterday, the rest should work just fine.

arinc9 commented:
The person I mentioned with Quectel RM520N-GL reports that the network works fine with your changes on umbim. They get high latency compared to using the QMI protocol, regardless of using ModemManager or umbim. Which makes me believe the issue lies on the cdc_mbim driver on the kernel. Therefore this is unrelated to this PR.

Let me know if you need any specifics about the network they're using.

By the way, the company I work with does a few things about cellular modems. We can run our own LTE or maybe even 5G network to try out all the different options with umbim. Keep in touch.

bmork commented:
> They get high latency compared to using the QMI protocol, regardless of using ModemManager or umbim. Which makes me believe the issue lies on the cdc_mbim driver on the kernel.

Might try to play with the tx_max and/or rx_max buffers to get better latency with MBIM.  Or simply set tx_timer_usecs to 0 and see how that goes.

Docs at https://www.kernel.org/doc/Documentation/ABI/testing/sysfs-class-net-cdc_ncm

Leo-PL commented:
> The person I mentioned with Quectel RM520N-GL reports that the network works fine with your changes on umbim. They get high latency compared to using the QMI protocol, regardless of using ModemManager or umbim. Which makes me believe the issue lies on the cdc_mbim driver on the kernel. Therefore this is unrelated to this PR.

Not surprised at all, but I still think this PR is worthy of merging. I think, that when I retest with IPv6, I'll mark it as ready again.
> 
> Let me know if you need any specifics about the network they're using.
I'm lucky enough to live in a country having a carrier having both IPv4 and IPv6, albeit on separate APNs. On QMI i was even able to use them both in dual-stack configuration, so I think I can cover every scenario.
> 
> By the way, the company I work with does a few things about cellular modems. We can run our own LTE or maybe even 5G network to try out all the different options with umbim. Keep in touch.
I'm planning on setting up LTE eNB in amateur band, using decomissioned hardware running in B40, which overlaps with one of amateur bands in Europe. Maybe we can join forces ;-)

arinc9 commented:
> Not surprised at all, but I still think this PR is worthy of merging. I think, that when I retest with IPv6, I'll mark it as ready again. It is big enough already.

Agreed.

> I'm lucky enough to live in a country having a carrier having both IPv4 and IPv6, albeit on separate APNs. On QMI i was even able to use them both in dual-stack configuration, so I think I can cover every scenario.

Perfect, I'll help in any way I can to get this merged.

> I'm planning on setting up LTE eNB in amateur band, using decomissioned hardware running in B40, which overlaps with one of amateur bands in Europe. Maybe we can join forces ;-)

I'd love to join forces. We'll be operating from Cyprus by summer, I join community events around Europe (been on Battlemesh last year). We'll meet at some point ;). I may also have other things we could work on together.

Leo-PL commented:
@arinc9 I retested the IPv6, and it works. The best I can do with IPv4v6 is to enforce autodetection of established bearer type - because the only mobile ISP in Poland with IPv6 doesn't officially do dual-stack. This autodetection actually works and avoids creation of sub-interfaces for the other bearer which isn't present.

Let's mark this as ready, then.

arinc9 commented:
> > They get high latency compared to using the QMI protocol, regardless of using ModemManager or umbim. Which makes me believe the issue lies on the cdc_mbim driver on the kernel.
> 
> Might try to play with the tx_max and/or rx_max buffers to get better latency with MBIM. Or simply set tx_timer_usecs to 0 and see how that goes.
> 
> Docs at https://www.kernel.org/doc/Documentation/ABI/testing/sysfs-class-net-cdc_ncm

They just reported the latency was extremely reduced! Thanks for pointing this out @bmork!

This PR works perfectly fine on our end.

Leo-PL commented:
Glad to hear that!
@arinc9 and @bmork, do you think it's worthy to add those sysfs knobs to the protocol handler, so they can be configured from UCI?

arinc9 commented:
If it's not going to be gruesome to implement, that'd be great!

Leo-PL commented:
Those knobs, however, are common to all NCM-based protocols - so I wonder how to handle that with NCM, QMI and MBIM protocol handlers without duplicating code.

Optimally, this should be handled by the protocol handler, because of possibility of hot-plugging the modems, which sometimes need hard reset, from my personal experience

dhewg commented:
Drive by comment: I know that netifd toggles some sysfs entries, maybe it'll fit there?

Leo-PL commented:
@arinc9 could you share the method you test the latency with? Just a standard ping or something more sophisticated?
I'd like to take a look at my MBIM and QMI modems as well.

arinc9 commented:
I haven't got the hardware myself, I'm assuming the person I'm in contact with just reads the RTT off speedtest.net.

ptpt52 commented:
How is the progress?

arinc9 commented:
@ptpt52 I don't think there's anyting left to work on here. It's waiting for someone to apply it. Maybe @hauke would be willing to do so.

Nowaker commented:
The endless saga continues. See here: https://github.com/openwrt/openwrt/pull/4405#issuecomment-960346606. It's been 2.5 years now since this was introduced: https://github.com/openwrt/openwrt/pull/3317. And 3 years since this: https://github.com/openwrt/openwrt/pull/2759. I'll post here again in 4 years to remind everybody about how bad the governance of this project is as I know for a fact this PR will remain unmerged.

hauke commented on package/network/utils/umbim/files/lib/netifd/proto/mbim.sh:
```diff
@@ -169,6 +169,8 @@ _proto_mbim_setup() {
 
 	echo "mbim[$$]" "Connected"
 
+	local zone="$(fw3 -q network "$interface" 2>/dev/null)"
```
Does this still work with fw4?

Leo-PL replied:
The same implementation is still used in other protocol handlers, and `/sbin/fw3` is just a symlink to to `/sbin/fw4`. Should I replace it? The same syntax with `fw4` works just as well as with `fw3`.

hauke replied:
ok thanks for confirming

hauke commented on package/network/utils/umbim/files/lib/netifd/proto/mbim.sh:
```diff
@@ -155,7 +155,7 @@ _proto_mbim_setup() {
 	local connect_state
 	echo "mbim[$$]" "Connect to network"
-	connect_state=$(umbim $DBG -n -t $tid -d $device connect "$req_pdptype:$apn" "$auth" "$username" "$password") || {
+	connect_state=$(umbim $DBG -n -t $tid -d $device connect "$req_pdptype$apn" "$auth" "$username" "$password") || {
```
How is this change related to the commit message? 
For me it looks like you should have done this in the previous commit already.

Leo-PL replied:
You're correct, that's a mistake when reordering commits. I'll push the fix when I figure out the MTU and MTU6 thing.

Leo-PL replied:
I decided not to wait with pushing this fix, while I try to figure out the MTU thing. I don't see any helpers for this in netifd repostiory either, so it's going to take some time.

hauke commented on package/network/utils/umbim/files/lib/netifd/proto/mbim.sh:
```diff
@@ -150,6 +153,7 @@ _proto_mbim_setup() {
 	tid=$((tid + 1))
 
 	pdptype=$(echo "$pdptype" | awk '{print tolower($0)}')
+	[ "$ipv6" = 0 ] && pdptype="ipv4"
```
Now `req_pdptype` will never be empty if IPv6 is not supported

Leo-PL replied:
This is expected, with IPv6 disabled, we can force IPv4 PDN connection in the next step anyway.

hauke commented on package/network/utils/umbim/files/lib/netifd/proto/mbim.sh:
```diff
@@ -201,6 +203,7 @@ _proto_mbim_setup() {
 			json_close_array
 			proto_add_dynamic_defaults
 			[ -n "$zone" ] && json_add_string zone "$zone"
+			[ -n "$ip4table" ] && json_add_string ip4table "$ip4table"
```
Does this still work with nftables?

Leo-PL replied:
This handles routing rules, and goes through netifd's abstraction, so I see no reason not to if it works for uqmi.

hauke commented on package/network/utils/umbim/files/lib/netifd/proto/mbim.sh:
```diff
@@ -294,6 +295,19 @@ _proto_mbim_setup() {
+	}
+	[ -n "$mtu" -a "$mtu" != 0 ] && {
+		echo Setting MTU of $ifname to $mtu
+		/sbin/ip link set dev $ifname mtu $mtu
```
Why do you set the mtu manually and do not create a json message?

netifd supports a mtu and a mtu6.

Leo-PL replied:
I based that off of implementation for uqmi. I'm trying to figure out where to add the `mtu` parameter to JSON, as for other examples I've seen it only in `tunnel` section, or so it seems - in conjunction with call to `proto_add_tunnel`.

Leo-PL replied:
Okay - I figured it out - this attribute is normally paired with `config device`, so I would have to update this with separate message, so netifd could pick it up.

hauke replied:
ok this is fine with me.

Leo-PL commented:
@hauke thanks for review. I'll perform some tests and perform updates needed for fw4 shortly, I believe that some compatibility is broken between fw3 and fw4 - this PR should be able work with both.

Leo-PL commented:
@hauke, I rebased and fixed the "$req_pdptype:" mistake. `fw3` and `fw4` are compatible in the matters discussed - so this works with both. And I'd like to keep "fw3" name in functions, unless there is a forward-compatibility symlink for fw4 provided by fw3. Everything still works as intended.

Leo-PL commented:
Hooray! Thanks for merging this!

Nowaker commented:
> https://github.com/openwrt/openwrt/pull/4405#issuecomment-1456742210

I'm happy I was wrong. Thank you @hauke for providing the needed guidance and support to help this years-long effort reach the finish line.
